### DISCUSSION TITLE
Leantime publish Dockerinstance in Subdirectory/Subfolder

### DISCUSSION THREAD

**Discussion by CrashOverride-lab** [Leantime/leantime#490] [Q&A]
Hi Guys,

i found this Project and Iám very impressed, so i will try out Leantime and show some other People the Project.

But i must publish Leantime behind a (Nginx-)Reverseproxy.

For Example: testnet.com/Leantime

@marcelfolaron writes, that this is in a Standard Installation from Version 2.1 possible. But how can i do that with the current Docker Container? Is there a Environment Variables for that available? If not, maybe someone has a Tipp or Configuration for me for my NGINX-Reverseproxy.

Thanks in advance.

With best regards

**Comment by marcelfolaron** [comment]
Hey,

Thanks for your feedback. We have a few people running it in nginx using docker. Here was an issue resolved on the docker repo:
https://github.com/Leantime/docker-leantime/issues/26

I think that is what you are looking for?
@LikoV2 Is that something you can help with? 

We should probably add that to the documentation

**Reply by CrashOverride-lab** [reply]
Hi,

honor where honor is due. :-).

First. Thank your for the Tipp. But my Setup is  a little bit different as the Link you posted (i need to publish it via a subdirectory). Or does LEAN_APP_URL support Subdirectorys? --> example: https://example.net/leantime?

**Reply by marcelfolaron** [reply]
Yes, the app url supports subfolder installs.

**Reply by CrashOverride-lab** [reply]
Hi Marcel,

sorry, but doesn´t work. A Dockercontainer of leantime with the latest Version and with the Parameter "-e LEAN_APP_URL=http://example.net/leantime" will bring back a 404 - Not found.

To verify it´s not my reverseproxyconfig i Opend: http://containerhost:Port/Leantime --> And here also i get a 404 not Found.
If i call http://containerhost:Port --> i get the Loginpage (raw text) because he couldn´t load Styles etc.

So if the Environment Variable really Supports Subdirectorys, then it seems (for me), that are some Bugs in there.

Or i missed something?

**Comment by marcelfolaron** [comment]
Hey, 

Without having seen your nginx config my first guess would be to check your document root:

```
set $base /var/www/<<subfolder for leantime (before /public)>>;
root $base/public;
```

**Reply by CrashOverride-lab** [reply]
Hi Marcel,

sorry, but i can´t follow you. The Test in my last post (directly access to the Container IP without NGINX) proofs, that it´s on the Container.

And there i work with the official docker container i can´t make direct changes in the  Container (because on a restart of the Container these Changes would be gone).

**Reply by marcelfolaron** [reply]
Correct, I see that Leantime is on the container. But for the subfolder installation to work property the document root of your nginx config needs to point to the public/ folder within the leantime directory. 

There are multiple ways to accomplish that. In any case you would have to provide your container a local config file for nginx.
Something similar to this should work:
https://stackoverflow.com/questions/42443468/nginx-location-configuration-subfolders

Someone else wrote an nginx config file for subfolder installs (@etm can you help out?):
https://github.com/Leantime/leantime/issues/330

Sorry my nginx knowledge is rather limited.

**Reply by AnthonyLELUYER** [reply]
Hello @CrashOverride-lab,

@marcelfolaron is right with the #330 nginx configuration, though not for the root subfolder, but for the Nginx rewrite capabilities.

When you reverse proxy with nginx, it's easier to rewrite than to configure the app to respond to subfolder. In most scenario, you will send a request to your nginx with http://example.com/leantime, and nginx will rewrite this URL to send to http://localhost:app_port.

It's a bit of work with nginx at first, but once you master the rewrite capabilities, you won't ever have any issue with any app that you will host.

**Reply by CrashOverride-lab** [reply]
Hi LikoV2,
Hi Marcel,

OK, so the LEAN_APP_URL Parameter cannot used for the Subdirectory Configuration for the Leantime Container. --> Glad, we could sort that out

I agree, it would be better if i could simply rewrite the Requests. But you see it for yourself in the amazing Post #330 from etm...

It is a hell of Configuration. And even if it works now... A new Version and a long Debugging will beginn again. Furthermore, people who are not so deeply involved in NGINX etc. (I count myself among them) have to be able to provide them easily, stably and quickly. And so far I haven't been able to find a NGINX-"master" Configuration anywhere on the Internet that just handles it.

Therefore, I and my colleagues and also larger companies take the "easy" way by forwarding the inquiries in NGINX using the location parameters.

Here is an excerpt from the NGINX config:

....

  location / app1 {
   proxy_pass http://10.0.0.1:8000;
   include /etc/nginx/conf.d/proxy.conf;
 }

  location / leantime {
   proxy_pass http://10.0.0.1:8003;
   include /etc/nginx/conf.d/proxy.conf;
 }

....

And there are 15 apps more. Provided easily and future-proof. But of course I like to learn from there. Can you give me an example NGINX configuration, where the rewrite is easy and future-proof and where I have the same advantages to deploy Apps in the same independent "Subdirectorys":

https://example.net/app1
https://example.net/app2
https://example.net/leantime

We all love this princip, because the user must only know 1 Hostname and the Programm-name which he wants use. And in the Root of Nginx (https://example.net) is a simple html document, which serves as a "table of contents". Perfect for all Users.

**Reply by marcelfolaron** [reply]
> OK, so the LEAN_APP_URL Parameter cannot used for the Subdirectory Configuration for the Leantime Container. --> Glad, we could sort that out

Not on its own, no. You will still have to set the APP URL once the nginx configs are complete.

Couldn't you just copy the config definition from #330 (everything within location /leantime/ {
and add that to the above excerpt? Leantime updates shouldn't cause issues with the re-directs (we fleshed that one out over the last year...)
@LikoV2 to confirm

**Reply by CrashOverride-lab** [reply]
Hi Guys,

so i tried as good as i could to adapt the Definitions form #330 and rewrite them for my Setup (a NGINx + a seperate Official Leantime-Docker-Container incl. LEAN_APP-URL Parameter).

Result: Not working. I get the same errors as with my configuration. 

Sad. Honestly i invested hours and iám currently very demotivated. 

But you 2 Guys --> Thank you very much. You tried to help a lot and answerd so quickly --> Its just great.

Maybe someone has a clear solution for the Standard  NGinx-Reversproxy + Official Leantime Docker Container Environment who simply need to published as a subsite. 

For any help: Thank you very much.

**Comment by marcelfolaron** [comment]
Sorry it didn't work!
I'll try to set up an nginx docker environment with subfolder installs this
weekend and see if I can find a solution.

**Reply by CrashOverride-lab** [reply]
Hey Marcel, thank you verry much. Maybe the simplest way would be implementing a Docker Environment Variable to publish the subfolder in the Leantime-Container (like Openproject has done with her Docker E.Variable: OPENPROJECT_RAILS__RELATIVE__URL__ROOT=/openproject)

**Comment by marcelfolaron** [comment]
That's a great idea. Let me check on that.
